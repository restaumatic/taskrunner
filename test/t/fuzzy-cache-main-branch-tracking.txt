# no toplevel
# s3

# We only download fuzzy cache if the _main branch commit_ changed since the last build.
# Why? In development, we don't want fuzzy cache to overwrite what we'we locally built.
# However, sometimes it's beneficial to incorporate upstream changes. A good approximation is:
# incorporate upstream changes when we get new commits on the main branch
# (assuming we're using a feature-branch workflow).

export TASKRUNNER_SAVE_REMOTE_CACHE=1
export TASKRUNNER_FALLBACK_BRANCHES="main"
export TASKRUNNER_MAIN_BRANCH="main"

git --bare init -q -b main upstream.git

mkdir a

(
  cd a
  echo foo > input.txt
  git init -q -b main
  git add input.txt
  git commit -q -m "commit1"
  
  git remote add origin ../upstream.git
  git push -q origin main
)

cp -r a b

go() {
  # Note: we want separate workdir for separate repos
  TASKRUNNER_STATE_DIRECTORY="$(pwd)" taskrunner -n mytask bash -e -c '
    snapshot input.txt --outputs output.txt --fuzzy-cache 
    echo -n "Expensive computation, "
    if [ -f output.txt ]; then
      echo "previous output is $(cat output.txt)"
    else
      echo "no previous output"
    fi
    rev input.txt > output.txt
  '
}

(
  cd a
  go
  echo "Output in a is: $(cat output.txt)"
)

(
  echo "*** 'bar' in b, should get fuzzy cache:"
  cd b
  export TASKRUNNER_SAVE_REMOTE_CACHE=0
  git checkout -b feature
  echo bar > input.txt
  go
  echo "Output in b is: $(cat output.txt)"
  git commit -q -m bar input.txt
)

(
  echo "*** 'baz' in b, should NOT get fuzzy cache:"
  cd b
  export TASKRUNNER_SAVE_REMOTE_CACHE=0
  echo baz > input.txt
  go
  echo "Output in b is: $(cat output.txt)"
  git commit -q -m bar input.txt
)

(
  echo "*** 'qux' in a, should upload cache"
  cd a
  echo qux > input.txt
  go
  echo "Output in a is: $(cat output.txt)"
  git commit -q -m qux input.txt
  git push -q origin main
)

(
  echo "*** merging and 'fud' in b, should get fuzzy cache"
  cd b
  git fetch -q origin main
  git merge -q -s ours origin/main

  export TASKRUNNER_SAVE_REMOTE_CACHE=0
  echo fud > input.txt
  go
  echo "Output in b is: $(cat output.txt)"
  git commit -q -m bar input.txt
)
