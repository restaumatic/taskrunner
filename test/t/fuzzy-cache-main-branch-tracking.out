-- output:
[mytask] info | Inputs changed, running task
[mytask] stdout | Expensive computation, no previous output
[mytask] info | success
Output in a is: oof
*** 'bar' in b, should get fuzzy cache:
Switched to a new branch 'feature'
[mytask] info | Inputs changed, running task
[mytask] info | Restored fuzzy cache from branch main, hash=6b36308659163b577fd1c832107dc46ca3aa659b
[mytask] stdout | Expensive computation, previous output is oof
[mytask] info | success
Output in b is: rab
*** 'baz' in b, should NOT get fuzzy cache:
[mytask] info | Inputs changed, running task
[mytask] stdout | Expensive computation, previous output is rab
[mytask] info | success
Output in b is: zab
*** 'qux' in a, should upload cache
[mytask] info | Inputs changed, running task
[mytask] stdout | Expensive computation, previous output is oof
[mytask] info | success
Output in a is: xuq
*** merging and 'fud' in b, should get fuzzy cache
[mytask] info | Inputs changed, running task
[mytask] info | Restored fuzzy cache from branch main, hash=22633d092c45c4ee1e6ca68e21dc09bdc038c0c8
[mytask] stdout | Expensive computation, previous output is xuq
[mytask] info | success
Output in b is: duf
